libraries/shared/storage/backfill.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/statediff"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
+)
+
+// IBackFiller is the backfilling interface
+type IBackFiller interface {
+	BackFill(bfa BackFillerArgs) (map[common.Hash][]utils.StorageDiff, error)
+}
+
+// BackFiller is the backfilling struct
+type BackFiller struct {
+	sdf fetcher.IStateDiffFetcher
+}
+
+// BackFillerArgs are used to pass configuration params to the backfiller
+type BackFillerArgs struct {
+	// mapping of hashed addresses to a list of the storage key hashes we want to collect at that address
+	WantedStorage map[common.Hash][]common.Hash
+	StartingBlock uint64
+	EndingBlock   uint64
+}
+
+// NewStorageBackFiller returns a IBackFiller
+func NewStorageBackFiller(bc fetcher.BatchClient) IBackFiller {
+	return &BackFiller{
+		sdf: fetcher.NewStateDiffFetcher(bc),
+	}
+}
+
+// BackFill uses the provided config to fetch and return the state diff at the specified blocknumber
+// StateDiffAt(ctx context.Context, blockNumber uint64) (*Payload, error)
+func (bf *BackFiller) BackFill(bfa BackFillerArgs) (map[common.Hash][]utils.StorageDiff, error) {
+	results := make(map[common.Hash][]utils.StorageDiff, len(bfa.WantedStorage))
+	if bfa.EndingBlock < bfa.StartingBlock {
+		return nil, errors.New("backfill: ending block number needs to be greater than starting block number")
+	}
+	blockHeights := make([]uint64, 0, bfa.EndingBlock-bfa.StartingBlock+1)
+	for i := bfa.StartingBlock; i <= bfa.EndingBlock; i++ {
+		blockHeights = append(blockHeights, i)
+	}
+	payloads, err := bf.sdf.FetchStateDiffsAt(blockHeights)
+	if err != nil {
+		return nil, err
+	}
+	for _, payload := range payloads {
+		block := new(types.Block)
+		blockDecodeErr := rlp.DecodeBytes(payload.BlockRlp, block)
+		if blockDecodeErr != nil {
+			return nil, blockDecodeErr
+		}
+		stateDiff := new(statediff.StateDiff)
+		stateDiffDecodeErr := rlp.DecodeBytes(payload.StateDiffRlp, stateDiff)
+		if stateDiffDecodeErr != nil {
+			return nil, stateDiffDecodeErr
+		}
+		accounts := utils.GetAccountsFromDiff(*stateDiff)
+		for _, account := range accounts {
+			if wantedHashedAddress(bfa.WantedStorage, common.BytesToHash(account.Key)) {
+				logrus.Trace(fmt.Sprintf("iterating through %d Storage values on account", len(account.Storage)))
+				for _, storage := range account.Storage {
+					if wantedHashedStorageKey(bfa.WantedStorage[common.BytesToHash(account.Key)], storage.Key) {
+						diff, formatErr := utils.FromGethStateDiff(account, stateDiff, storage)
+						logrus.Trace("adding storage diff to out channel",
+							"keccak of address: ", diff.HashedAddress.Hex(),
+							"block height: ", diff.BlockHeight,
+							"storage key: ", diff.StorageKey.Hex(),
+							"storage value: ", diff.StorageValue.Hex())
+						if formatErr != nil {
```

Comment:
Do we potentially want to do this check before we log the trace? Thinking we could maybe hit null-pointer panics on the call to `.Hex()` if there was an error and fields on the diff are `nil`

Reply:
Good catch, that could lead to a panic!